src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFindOptimizations.cs

```diff
@@ -225,6 +239,62 @@ public RegexFindOptimizations(RegexNode root, RegexOptions options)
         /// <summary>When in literal after set loop node, gets the literal to search for and the RegexNode representing the leading loop.</summary>
         public (RegexNode LoopNode, (char Char, string? String, char[]? Chars) Literal)? LiteralAfterLoop { get; }
 
+        /// <summary>Analyzes a list of fixed-distance sets to extract a case-sensitive string at a fixed distance.</summary>
+        private static (string String, int Distance)? FindFixedDistanceString(List<(char[]? Chars, string Set, int Distance)> fixedDistanceSets)
+        {
+            (string String, int Distance)? best = null;
+
+            // A result string must be at least two characters in length; therefore we require at least that many sets.
+            if (fixedDistanceSets.Count >= 2)
+            {
+                // We're walking the sets from beginning to end, so we need them sorted by distance.
+                fixedDistanceSets.Sort((s1, s2) => s1.Distance.CompareTo(s2.Distance));
+
+                Span<char> scratch = stackalloc char[64];
+                var vsb = new ValueStringBuilder(scratch);
+
+                // Looking for strings of length >= 2
+                int start = -1;
+                for (int i = 0; i < fixedDistanceSets.Count + 1; i++)
+                {
+                    char[]? chars = i < fixedDistanceSets.Count ? fixedDistanceSets[i].Chars : null;
+                    bool invalidChars = chars is not { Length: 1 };
+
+                    // If the current set ends a sequence (or we've walked off the end), see whether
+                    // what we've gathered constitues a valid string, and if it's better than the
+                    // best we've already seen, store it.  Regardless, reset the sequence in order
+                    // to continue analyzing.
+                    if (invalidChars ||
+                        (i > 0 && fixedDistanceSets[i].Distance != fixedDistanceSets[i - 1].Distance + 1))
+                    {
+                        if (start != -1 && i - start >= (best is null ? 2 : best.Value.String.Length))
+                        {
+                            best = (vsb.ToString(), fixedDistanceSets[start].Distance);
+                        }
+
+                        vsb = new ValueStringBuilder(scratch);
+                        start = -1;
+                        if (invalidChars)
+                        {
+                            continue;
+                        }
+                    }
+
+                    if (start == -1)
+                    {
+                        start = i;
+                    }
+
+                    Debug.Assert(chars is { Length: 1 });
+                    vsb.Append(chars[0]);
```

Comment:
Is it possible to run out of space in the 64 char array? Should we be guarding against that?

Reply:
That's just the initial seed space.  If it runs out of room ValueStringBuilder will expand with ArrayPool memory.  We can pass an empty span and it would still work, we just start with enough stack space to cover the overwhelming majority of cases.

Reply:
oh I wasn't aware of that, thanks for explaining.